base/abstractarray.jl

```diff
@@ -2313,7 +2317,7 @@ function _typed_hvncat(::Type{T}, shape::Tuple{Vararg{Tuple, N}}, row_first::Boo
 
             wasstartblock = blockcounts[d] == 1 # remember for next dimension
 
-            isendblock = blockcounts[d] == shape[d][shapepos[d]]
+            isendblock = blockcounts[d] == shape[d > N ? end : d][shapepos[d]]
```

Comment:
And while I'm here, I find this slightly better since I had to double-check that `N == length(shape)`.

```suggestion
            isendblock = blockcounts[d] == shape[min(end, d)][shapepos[d]]
```